content/docs/rfcs/47/README.md

```diff
@@ -0,0 +1,94 @@
+
+## The CLIENT Socket Type
+
+The CLIENT socket type talks to a set of anonymous peers, sending and receiving messages using round-robin algorithms. It is reliable, insofar as it does not drop messages. 
+
+General behavior:
+
+* MAY be connected to any number of SERVER peers, and MAY both send and receive messages.
+* SHALL not filter or modify outgoing or incoming messages in any way.
+* SHALL maintain a double queue for each connected peer, allowing outgoing and incoming messages to be queued independently.
+* SHALL create a double queue when initiating an outgoing connection to a peer, and SHALL maintain the double queue whether or not the connection is established.
+* SHALL create a double queue when a peer connects to it. If this peer disconnects, the DEALER socket SHALL destroy its double queue and SHALL discard any messages it contains.
+* SHOULD constrain incoming and outgoing queue sizes to a runtime-configurable limit.
+
+For processing outgoing messages:
+
+* SHALL consider a peer as available only when it has a outgoing queue that is not full.
+* SHALL route outgoing messages to available peers using a round-robin strategy.
+* SHALL block on sending, or return a suitable error, when it has no available peers.
+* SHALL not accept further messages when it has no available peers.
+* SHALL NOT discard messages that it cannot queue.
+* MUST NOT send multipart messages.
+
+For processing incoming messages:
+
+* SHALL receive incoming messages from its peers using a fair-queuing strategy.
+* SHALL deliver these to its calling application.
+* MUST discard any part of a multipart message, MAY disconnect a peer that is sending multipart messages.
+
+## The SERVER Socket Type
+
+The SERVER socket type talks to a set of peers, using an explicit routing-id so that each outgoing message is sent to a specific peer connection.
+
+General behavior:
+
+* MAY be connected to any number of CLIENT peers, and MAY both send and receive messages.
+* SHALL maintain a double queue for each connected peer, allowing outgoing and incoming messages to be queued independently.
+* SHALL create a double queue when initiating an outgoing connection to a peer, and SHALL maintain the double queue whether or not the connection is established.
+* SHALL create a double queue when a peer connects to it. If this peer disconnects, the ROUTER socket SHALL destroy its double queue and SHALL discard any messages it contains.
+* SHALL assign a routing-id to each double queue.
+* Routing-id SHALL be 4-bytes integer.
+* SHALL NOT allow the peer to specify its routing id explicitly.
+* SHOULD constrain incoming and outgoing queue sizes to a runtime-configurable limit.
+
+For processing incoming messages:
+
+* SHALL receive incoming messages from its peers using a fair-queuing strategy.
+* SHALL attach the routing-id of double queue to the message.
```

Comment:
Attach where or how?

Reply:
Stealing from REQREP:
SHALL prefix each message with a frame containing the *routing id* of the originating double queue.

Reply:
But it is not prefixed. If it was prefixed it was not thread-safe.
It is attached to the message, how is more implementation-dependent.

Reply:
I added a clarification to the ZMTP 3.1 RFC